Add DevTools <-> Server API to allow server to have DevTools re-connect to new VM services

DanTup opened the pull request:
@jacob314 @devoncarew before I go too far with this, I want to check I'm going in the right direction...

This adds a basic SSE connection between server+client, and the server will track the clients that connect to it, expecting to find ones that are not longer connected to VM services to reuse them instead of launching new browsers.

@grouma FYI - you might have some comments on the SSE stuff, for ex:

- is using `connections.rest.listen` a bit weird? The example uses `await for` but I have other code running here to handle the other DevTools serving so don't want to end up waiting around and I wasn't sure of another way to subscribe to the stream).

- what's the best way to tell if a client is still alive? I can't see any obvious close (or heartbeat) events on the `SseConnection`.


Something else we need to think about is how to handle when there's no server (eg. when  you want to run/debug DevTools, you need to run without the server - since the server currently serves a built release version). I currently just have a try/catch around the connection and print a message - is this reasonable, and do we intend to always ensure DevTools can run without the server via `webdev serve`?

- [x] Needs tests
- [ ] Running on a random port means new notifications permission requests each time (may need fixing in VS Code)
- [x] ~Need to not reload the page when reconnecting, since it prevents using a notification to tell the user (if we send the notification before the reload, clicking it will open a new tab)~ worked around and opened https://github.com/flutter/devtools/issues/989
- [x] ~Need to add some version checking/error handling to ensure there aren't issues if server/client versions aren't matched as this rolls out (eg. when landing this, the pubspec for devtools will still reference a version of server without these changes)~ Not an issue since we pin exact versions
- [x] Don't connect to API for DDC (debug) builds
- [x] Notifications will be spammy if trying to reconnect DevTools automatically on new debug sessions - pass a flag through from the editor to only notify on user actions?
- [x] Fix 5s timeout in tests to instead receive a notification whenever a new client is spawned or one reused
- [ ] Ensure server tests are enabled after we publish a server version with these changes
- [ ] Ensure notifications are hidden automatically if DevTools gets focus
- [ ] Test if `window.focus()` is better than notifications

grouma commented:
> - is using `connections.rest.listen` a bit weird? The example uses `await for` but I have other code running here to handle the other DevTools serving so don't want to end up waiting around and I wasn't sure of another way to subscribe to the stream).

I don't see anything wrong with that, it's a part of the StreamQueue API.

- what's the best way to tell if a client is still alive? I can't see any obvious close (or heartbeat) events on the `SseConnection`.

You can listen to `connection.sink.done` or `connection.stream.listen(onDone:)`

jacob314 commented on packages/devtools/lib/src/main.dart:
```diff
@@ -136,6 +138,19 @@ class PerfToolFramework extends Framework {
+        channel.sink.add('Test message from client');
+      });
+    } catch (e) {
+      print('Failed to connect to SSE API: $e');
```
Bonus points for logging this as an error in release (dart2js) builds and only displaying it as a print statement warning is debug (ddk) builds. If debug build I expect the endpoint won't typically be available.

DanTup replied:
Is there any way to tell which version we were built in?

jacob314 replied:
yes. replied to the overall thread with an example of detecting whether asserts are enabled.

jacob314 commented on packages/devtools_server/lib/src/server.dart:
```diff
@@ -24,6 +26,8 @@ const launchDevToolsService = 'launchDevTools';
 
 const errorLaunchingBrowserCode = 500;
 
+final ClientManager clients = new ClientManager();
```
nit: run dartfmt --fix on this file to eliminate uses of new here and elsewhere.

DanTup replied:
Should/can we enable the `unnecessary_new` lint so we can see these in the editor?

jacob314 replied:
make it so!

DanTup replied:
https://github.com/flutter/devtools/pull/998

FWIW, when I ran `dartfmt --fix` there was no output (and no changed files) after many minutes, so I terminated it and did it manually... is it expected to be really slow, or is that a bug?

jacob314 replied:
sorry I left out a few of the arguments you need to pass to `dartfmt fix`.

`dartfmt --fix  -w directory/to/run/on/`

jacob314 commented on packages/devtools_server/lib/src/server.dart:
```diff
@@ -88,13 +92,18 @@ void serveDevTools({
         // so it's relative to packHandler's root.
         ? packHandler(request.change(path: 'packages'))
-        : buildHandler(request);
+        : request.url.path.startsWith('api/sse')
```
I would move this logic into a few if statements each of which ends in a return. It is now too complex to be using conditionals like this).

DanTup replied:
👍

jacob314 commented on packages/devtools_server/lib/src/server.dart:
```diff
@@ -203,6 +212,15 @@ Future<void> registerLaunchDevToolsService(
+        final reusableClient = await clients.findReusableClient();
+        if (reusableClient != null) {
+          await reusableClient.connectToVmService(vmServiceUri);
+          // reusableClient.bringToFront() / Notify / whatever...
```
the bringToFront part is the million £ question. I expect notify is the best we will be able to do.

DanTup replied:
I went with notifications, though I have a note to do some testing of how `window.focus()` behaves these days, in case that works well.

jacob314 commented on packages/devtools_server/lib/src/client_manager.dart:
```diff
@@ -0,0 +1,35 @@
+  /// and it disconnected, then started debugging again, we want to reuse
+  /// the open DevTools window).
+  Future<DevToolsClient> findReusableClient() async {
+    // TODO(dantup):
```
I would suggest have the clients send messages when a client becomes reusable and when it stops being reusable.
For example, a client could be reusable and then the user could connect to another VM Service or drag in a snapshot of a CPU Profile in which case it is not reusable.

DanTup replied:
Yeah, my plan here was to add a boolean in `DevToolsClient` that tracks whether that client is connected to a VM (and have DevTools ping on connect/disconnect), then have this code just find the first one that's not.

jacob314 replied:
> * It required exposing the list of clients in the server API (this might not be a bad thing - maybe there's some future value in editors being able to enumerate the list of open DevTools tabs for the server that it owns - for now the reason is just to help test this is working - we could prefix it with an underscore or something if we'd prefer to consider it "private")

That is fine.

DanTup commented:
Pushed some updates. This basically works from VS Code now, but it has a few things that need resolving:

- We need to show notification permission requests at startup (if we leave it until the first time we try to show a notification, it could be in the background and the user will wonder why the DevTools button in the editor isn't doing anything)
- Spawning on a random port each time means notification permissions each time - we need to find a better way to prefer the same port, and increase if it's not available (or something)
- Reconnecting after a device disconnected doesn't seem trivial, so for now I'm reloading the page - however this means I can't send notifications (if I send them before the reload, clicking them will load a new browser tab since the URL has changed)

Some other things of note:

I also had to stash the VM Service URI on VmServiceWrapper so we can send it to the server (since as well as reusing disconnected tabs, we want to support just notifying from an existing connection if the user hits the DevTools button again in their editor when it's already up and running) - I don't know if the way I did this was the best way.

The JSON building/handling is currently just hand-written - it'd be nice if this was generated so we have a strong interface between client/server. How worthwhile this is may depend on how much it'll be used (we can always do it if/when we build on it).

It still needs tests (I'll add them once I'm sure the implementation isn't going to change significantly).

DanTup commented:
I made a start on testing.. I've pushed a basic test that DevTools connects back to the SSE endpoint, however:

- It required exposing the list of clients in the server API (this might not be a bad thing - maybe there's some future value in editors being able to enumerate the list of open DevTools tabs for the server that it owns - for now the reason is just to help test this is working - we could prefix it with an underscore or something if we'd prefer to consider it "private")
- The test won't work on CI yet, because in the repository devtools references devtools_server from pub - and that version doesn't have the SSE endpoint. We can run the test locally and enable it after the next release - however we probably want to think a little more about this... We could rig the pubspec.yaml during CI runs, but really we may  want to be sure that any changes work with *both* the repo version of the server and the live version of the server, to avoid coupling the releases too tightly?

DanTup commented:
Pushed more tests. One of them isn't done yet because I need a change in `browser_launch` (PR submitted), but the others work - however they're marked `skip` because the tests currently run with the Pub-published version (detailed above).

VS Code re-uses the same port number within a session, but does pass `--port=0` for the first run each session - this means lots of Notification permissions prompts. I don't know what a good fix is there (we can hard-code a port, but if it's taken then we just fail). It'd be nice if we could tell Dart we want the first available port starting at `x` rather than just a random one, to try and use a narrower range 🤔

Finally, when you run and the SSE endpoint is a 404 (for ex. running DDC), you get 404s printed to the browser console. I don't know if there's a way to prevent this since they look like normal network requests (@grouma may know?) but if we can detect running under DDC, we could at least skip the attempts to connect to the API to silence them. I'm not sure if we can detect running under DDC/`webdev serve` though. We could add another querystring param to the release version, but not sure if that's the best way.

jacob314 commented:
We'll never release a production build with asserts on so checking for the presence of asserts is a fine way to detect debug ddc vs production builds. For example:

```dart
bool isDebugBuild() {
  bool debugBuild = false;
  assert((() { debugBuild = true; return true; })());
  return debugBuild;
}
```

jacob314 commented on packages/devtools/lib/src/server_api_client.dart:
```diff
@@ -0,0 +1,80 @@
+    window.location
+        .replace(uri.replace(queryParameters: newUriParams).toString());
+
+    // TODO(dantup): Could we make devtools able to reconnect more gracefully?
```
nit: make this todo also reference 989 or combine the todos.

jacob314 commented on packages/devtools/lib/src/server_api_client.dart:
```diff
@@ -0,0 +1,80 @@
+      return;
+    }
+
+    Notification('Dart DevTools',
```
We should not notify if the page is already visible as that will be really annoying.
I think the page visibility APIs will be able to do a good job of this but I haven't verified.
https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API

DanTup replied:
The Visibility API is only for when the tab is visible or not. If you havea browser with DevTools as the active tab, but it's hidden (eg. you only have one screen and have VS Code covering the whole thing) it would appear as active.

For the case where the user has explicitly pressed a button in VS Code, I think the notification is ok (since if you had DevTools visible you'd probably just click over on the window instead of pressing the button in VS Code), however it would suck if we add an option to automatically launch it.

Maybe we should pass a flag through from the editor for whether to notify? If it's a direct user action, we do - and if it's not, then we're silent?

jacob314 replied:
Having a flag in the API indicating whether to notify sounds reasonable. Can we close the notification as soon as the user navigates to devtools? That would probably be a fine user experience.

DanTup replied:
I've added a flag so notifications are opt-in by the editor launching DevTools, and made VS Code (currently on a branch) only set it when there was a user interaction (eg. if you use the new "open DevTools automatically when starting a debug session", it will not generate notifications).

I'll make sure the notification closes when DevTools gets focus too.

jacob314 commented on packages/devtools/test/integration_tests/server_test.dart:
```diff
@@ -12,10 +13,13 @@ import '../support/cli_test_driver.dart';
-  DevToolsServerDriver server;
+CliAppFixture appFixture;
+DevToolsServerDriver server;
+final Map<String, Completer<Map<String, dynamic>>> completers =
```
remove the massive generic types from the RHS.

jacob314 replied:
> 2. VS Code currently passed `--port=0` to get a random port, to avoid issues with the port already being in use. Chrome saves notification permissions based on host/port which means the user will be prompted to approve notifications each time it spawns on a port they haven't accepted before. Switching to a hard-coded port would avoid this, but may introduce other issues.

We should switch to a hard coded port with devtools falling back to increasing port #s. We've been too aggressive about using a random port #. The other option long term is to actually serve devtools from a website which would fix the issue.

DanTup replied:
I've opened https://github.com/flutter/devtools/issues/1037 to try ascending ports - and I'll work on that and land it before this one lands.

jacob314 commented on packages/devtools/test/integration_tests/server_test.dart:
```diff
@@ -42,34 +59,180 @@ void main() {
+    expect(serverResponse['clients'][0]['hasConnection'], isTrue);
+    expect(serverResponse['clients'][0]['vmServiceUri'],
+        equals(appFixture.serviceUri.toString()));
+    // TODO(dantup): This test will fail until the devtools pubspec.yaml
```
add a bug with this TODO or before landing break it up so you commit this integration test after committing the other changes on this cl.

DanTup replied:
It's not just landing the change, in order to unskip this we need to push a new version of the server to pub, and then update the version number in devtools/pubspec.yaml. This is because the tests run having pulled devtools_server from pub, and not the version in the repo.

We could rig the tests to use the local version, but I think there's also value in testing against the pub version (after we've pushed this first version) to ensure we don't accidentally publish a DevTools that breaks against the live server (because it was only tested against unpublished code).

We could potentially run the tests twice, but I don't know how simple that'll be. Any preferences?

(If we land this as-is, I'll open an issue to unskip these after we next publish the server - it may be worthwhile doing dev builds of each just for some extra validation).

jacob314 commented on packages/devtools/test/integration_tests/server_test.dart:
```diff
@@ -42,34 +59,180 @@ void main() {
+    // was not a second client added.
+    await appFixture.serviceConnection.callMethod(
+        registeredServices['launchDevTools'],
+        args: {'reuseWindows': true});
```
nit: add trailing comma to make this code format better.

jacob314 commented on packages/devtools/test/integration_tests/server_test.dart:
```diff
@@ -42,34 +59,180 @@ void main() {
+    // This delay needs to be long enough to allow DevTools to load and connect
+    // back. If it's too short, the test may pass only because we didn't wait
+    // long enough.
+    await Future.delayed(const Duration(seconds: 5));
```
this timeout scares me. We could get flakes on slow machines and we will be slowing down everyone's test runs by 5 seconds on fast machines. Can we instead wait for an event from the client that devtools is running?

DanTup replied:
It wouldn't be flaky, as it'd "fail" green if the time was too short, but I agree this feels weird.

I'll add a notification each time a client connects, and we can wait on that and then ensure there's still only one connected client.

DanTup replied:
I've added a new event (`client.launch`) that the server emits whenever a client is launched now - it includes booleans for whether it reused an existing window and/or showed a notification. For now this is only used by the tests, but it may be handy for the editors to  consume in future.

jacob314 commented on packages/devtools/test/integration_tests/server_test.dart:
```diff
@@ -42,34 +59,180 @@ void main() {
+    // long enough.
+    await Future.delayed(const Duration(seconds: 5));
+
+    // Ensure there's still only one connection (eg. we didn't spawn a new one
```
nice!

jacob314 commented on packages/devtools/test/integration_tests/server_test.dart:
```diff
@@ -42,34 +59,180 @@ void main() {
+    },
+    timeout: const Duration(seconds: 10),
+    timeoutMessage: 'Server did not return any known clients',
+    delay: const Duration(seconds: 1),
```
using this sort of pattern instead of the 5 second timeout used earlier would also be fine.

DanTup replied:
This one is easy because we're waiting for something to happen, but the other one has no visible effects (because it would reuse an existing connection, and the client list would look the same). I think an event will work though, so I'll do that.

DanTup commented:
Pushed changes addressing most of the comments, though still some outstanding (I've responded to the comments and added them to the checklist in the top comment - it'll be next  week before I get to resolve these now).

grouma commented:
> Finally, when you run and the SSE endpoint is a 404 (for ex. running DDC), you get 404s printed to the browser console. I don't know if there's a way to prevent this since they look like normal network requests (@grouma may know?) but if we can detect running under DDC, we could at least skip the attempts to connect to the API to silence them. I'm not sure if we can detect running under DDC/webdev serve though. We could add another querystring param to the release version, but not sure if that's the best way.

Can you give examples here? package:sse doesn't have any printing logic so it shouldn't be coming from that. Maybe you have a top level error handler that is printing?

DanTup commented:
> Can you give examples here? package:sse doesn't have any printing logic so it shouldn't be coming from that. Maybe you have a top level error handler that is printing?

I think it's the browser printing them because it sees HTTP requests that are 404s (they appear as red bars in Chrome). I've resolved this a different way now though (don't try to set up the API when running a debug build, since the server will only be used for release builds) so I don't think it's a problem.

DanTup commented:
@jacob314 I think this is mostly there now, with two outstanding issues I'm not sure of the best solutions for:

1. Tests are currently marked skip, because they require the server code from the repo, but the tests run with the server from pub
2. VS Code currently passed `--port=0` to get a random port, to avoid issues with the port already being in use. Chrome saves notification permissions based on host/port which means the user will be prompted to approve notifications each time it spawns on a port they haven't accepted before. Switching to a hard-coded port would avoid this, but may introduce other issues.

The first is related to our discussions yesterday about versioning dependent packages which I think you're thinking more about.

For the second, I'm wondering if we should have an option where the `--port` passed is just a "starting port" and will be increased by up to x (using try/catch) until it gets a port. It means the client isn't guaranteed to get the port they want, but they were be consistent (within a small range) to increase the chances of overlapping and not presenting duplicate notifications permission dialogs. WDYT?

jacob314 commented:
> The first is related to our discussions yesterday about versioning dependent packages which I think you're thinking more about.

For devtools and devtools server I think the right solution is to check in the devtools code with a valid pubspec version combined with a dependency override for devtools server.
We will then change change the publish script to remove the override and publish both version s of the packages.
What is required for this to work is as soon as someone changes devtools or devtools server they need to update the version in the pubspec and update the changelog if they haven't yet. Go ahead and add the dev dependency override as part of your change to test it.

jacob314 commented on packages/devtools/lib/src/main.dart:
```diff
@@ -136,6 +140,44 @@ class PerfToolFramework extends Framework {
     screensReady.future.then(app.devToolsReady);
   }
 
+  void initSseConnection() {
```
this name is too generic. We use SSE to connect to the VMService as well.
Maybe make this method:
`initDevtoolsServerConnection()` instead

jacob314 commented on packages/devtools/lib/src/main.dart:
```diff
@@ -136,6 +140,44 @@ class PerfToolFramework extends Framework {
+  void initSseConnection() {
+    // When running the debug DDC Build, the server won't be running so we
+    // can't connect to its API (for now at least, the API is optional).
+    if (isDebugBuild()) {
```
we should provide a way to debug the server workflows in debug mode. File a bug for us to configure our environment to support that.

DanTup replied:
I've opened https://github.com/flutter/devtools/issues/1036. It's not terribly bad as-is, my VS Code launch config automatically does a release build before running - so you can just press F5 and it'll do a release build, run the server, spawn the browser - ofc that's slower than if we were using the DDC one.

jacob314 commented on packages/devtools/lib/src/main.dart:
```diff
@@ -136,6 +140,44 @@ class PerfToolFramework extends Framework {
+    }
+
+    try {
+      apiClient = ServerApiClient();
```
apiClient is also a bit too generic a name. every client has an api. Give this a name that indicates it is the DevtoolsServerClient.

DanTup replied:
"ServerClient" looks a bit confusing, so I just named it `devToolsServer`, so calls look like this:

```dart
devToolsServer.notifyConnected(serviceManager.service.connectedUri);
```

Seem reasonable?

jacob314 commented on packages/devtools/lib/src/main.dart:
```diff
@@ -136,6 +140,44 @@ class PerfToolFramework extends Framework {
+        }
+      });
+    } catch (e) {
+      print('Failed to connect to SSE API: $e');
```
use the new logging API kenzie added (after you merge)

jacob314 commented on packages/devtools/lib/src/server_api_client.dart:
```diff
@@ -0,0 +1,87 @@
+
+import 'package:sse/client/sse_client.dart';
+
+class ServerApiClient {
```
rename this to 
DevtoolsServerClient

jacob314 commented on packages/devtools/lib/src/server_api_client.dart:
```diff
@@ -0,0 +1,87 @@
+import 'package:sse/client/sse_client.dart';
+
+class ServerApiClient {
+  ServerApiClient() : _channel = SseClient('/api/sse') {
```
to enable using this in a debug workflow we should not hardcode it to assume the SseClient is on the same hostname as the app.
If you add a query parameter to devtools passing in the hostname of the server then it will be easier to test this with debug builds if people need to. That will matter more when we make the server vend the analysis server protocol. We also need to pass in a security token like what the Dart VMService does to the SseClient to avoid a security vulnerability.

jacob314 commented on packages/devtools/lib/src/server_api_client.dart:
```diff
@@ -0,0 +1,87 @@
+            notify();
+            return;
+          default:
+            print('Unknown request ${request.method} from server');
```
use logging package.

jacob314 commented on packages/devtools/lib/src/server_api_client.dart:
```diff
@@ -0,0 +1,87 @@
+    _channel.stream.listen((msg) {
+      try {
+        final request = jsonDecode(msg);
+        switch (request['method']) {
```
nit:
add assert(request['method') != null)

jacob314 commented on packages/devtools/lib/src/server_api_client.dart:
```diff
@@ -0,0 +1,87 @@
+  final SseClient _channel;
+
+  int _nextRequestId = 0;
+  void _send(String method, [dynamic params]) {
```
can you make params a `Map<String, dynamic>`?

DanTup replied:
Yep!

DanTup commented:
I've rebased this on the latest changes. There were conflicts around a few changes:

1. The code that built the handlers was extracted to a new file while I added a new one. Moving my code was easy, however whatever is passing its own handler in is probably not going to handle the SSE API connections. @DaveShuckerow FYI - I don't know what the consuming code looks like, but you can see the change in the `defaultHandler` function above. It registers an API that's used for DevTools to communicate with the server. You may need to reproduce this elsewhere.

2. The file moves all worked fine (yay Git!) except a new file remained in the `devtools` folder that needed to be in `devtools_app`, so that was an easy fix.

I pushed a copy of the original branch prior to rebasing https://github.com/flutter/devtools/compare/master...DanTup:server-api-backup so I could easily compare them to ensure I didn't trash anything whole resolving the conflicts.

I want to do a little testing after these changes (and ensure the tests all pass - since they're not running on CI because of complications mentioned above) before landing - but @jacob314 lmk if you're still happy to land this as-is if I don't hit any issues (there are some things we need to tidy up before releasing, however this work is mostly opt-in so I *think* we could ship it and then enable it in VS Code later).

jacob314 commented:
![lgtm](https://i.imgur.com/JgNRalI.png)
This is good enough to land as is and cleanup after.

DanTup commented on packages/devtools_app/lib/src/main.dart:
```diff
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'package:devtools_app/src/server_api_client.dart';
```
Oops, good catch - I used the quick-fix to re-insert these after rebasing over the move and didn't pay attention to which one I selected 🙈 Both fixed!

Looks like we have the `avoid_relative_lib_imports` lint enabled in analysis_options, which seems to want the *opposite* of this? (yet for reasons I can't explain, it's not flagging anything up?)

DanTup commented:
Pushed one more change - 42dfb6628e229881e0ccfa2b7fe4aa08bf29febb this puts the initial notifications permission request behind a flag, so we'll only request permissions if the editor starting the server tells us to (making it opt-in, so landing this shouldn't change any behaviour - besides the SSE connection being opened).

That said, I'm now wondering if we should change this flag to be `--enable-api` instead, and put the whole SSE API behind it instead of just notifications. @jacob314 WDYT?

DanTup commented:
@DaveShuckerow Great, thanks for confirming!

I'll wait for @jacob314 to confirm he's happy with the final permissions changes (or if we should put the whole SSE behind a flag) before landing.

jacob314 commented:
Still LGTM. Please ping me if you are ever waiting more than a few hours for a reply on a CL.
![lgtm_dog](https://user-images.githubusercontent.com/1226812/65183291-08031200-da18-11e9-8c20-061ddd21a334.jpeg)

DanTup commented:
Will do! I was out last week so I wasn't rushing to land this before I was back in case it broke anything. I've rebased this (conflicts were just in imports) and will give it another test then land it. Thanks!

DanTup commented:
Looks good locally - build failure looks like same as on master, so I'll re-run them once that's resolved.

DanTup commented:
It went green, though had some conflicts with @DaveShuckerow's API. I've fixed these up, though we both used `/api` so the handler now looks like this:

```dart
if (request.url.path.startsWith('api/sse')) {
  return sseHandler.handler(request);
}

// The API handler takes all other calls to api/.
if (ServerApi.canHandle(request)) {
  return ServerApi.handle(request);
}
```

Does this seem reasonable, or should I avoid eating into `ServerApi`'s space?

DaveShuckerow commented:
Can you make sseHandler part of the ServerApi class?

That will make it easiest to handle in bazel environments too.

DanTup commented:
> Can you make sseHandler part of the ServerApi class?

Do you mean just move the `SseHandler` inside the ServerApi class? (we can't change much about it, as the whole `SseHandler` class comes from the SSE package). We do have it add connections to a client manager so we can interact with them, so we'd need to also pass that into ServerApi:

```dart
  final sseHandler = SseHandler(Uri.parse('/api/sse'))
    ..connections.rest.listen(clients.acceptClient);
```

I don't know anything about Bazel to know what the implications are of any of this though.

DaveShuckerow commented on packages/devtools_server/lib/src/handlers.dart:
```diff
@@ -60,7 +74,8 @@ class ServerApi {
+  static FutureOr<shelf.Response> handle(shelf.Request request,
+      [ServerApi api]) {
     api ??= ServerApi();
     switch (request.url.path) {
```
@DanTup 

Can you move the sseHandler logic here?

```dart
class ServerApi {
  final SseHandler sseHandler = SseHandler(Uri.parse('/api/sse'))
    ..connections.rest.listen(clients.acceptClient);

  handle(...) {
    switch ... {
    case 'api/sse':
      return sseHandler.handler(request);
  ...
    }
  }
}
```

DaveShuckerow replied:
I realize this might mean constructing the ServerApi instance before line 40, depending on what sseHandler wants.

DaveShuckerow commented:
> > Can you make sseHandler part of the ServerApi class?
> 
> Do you mean just move the `SseHandler` inside the ServerApi class? (we can't change much about it, as the whole `SseHandler` class comes from the SSE package). We do have it add connections to a client manager so we can interact with them, so we'd need to also pass that into ServerApi:
> 
> ```dart
>   final sseHandler = SseHandler(Uri.parse('/api/sse'))
>     ..connections.rest.listen(clients.acceptClient);
> ```
> 
> I don't know anything about Bazel to know what the implications are of any of this though.

I left comments inline on what we can do here.  The Bazel version will create a ServerApi subclass to override some of the endpoints.  It also defines a different default handler to handle static file serving differently.

If the suggestions I left work, then that LGTM.  If not, this isn't important enough to block you any longer, so feel free to submit and I can refactor if need be.

DanTup commented:
We can't move the SseHandler construction into ServerApi without somehow giving it access to `clients`. I tried passing it into `handle` and that worked, however if you're subclassing ServerApi somewhere else, I don't think that'll work either (since you won't pass in the SseHandler).

>  feel free to submit and I can refactor if need be.

Ok, I think that's probably easiest - you have a better understanding of the other side of this, and the longer this remains in a PR, the more I have to keep rebasing and re-testing it (there are new conflicts again since I rebased last night :-))

So, I'll fix the latest conflicts and if the build is green land this and you're free to make changes (there's nothing outstanding for me to do here for now, so I won't be making simultaneous changes). If it seems like this may have affected anything, feel free to revert it if I'm not around.
